### DISCUSSION TITLE
Multi-floppy installations

### DISCUSSION THREAD

**Discussion by nelgin** [dosemu2/dosemu2#2241] [General]
Hi all,

I'm trying to install a DOS package that comes with 4 floppy images. It will only install from a floppy drive. Once I run the install it'll ask for the 2nd floppy. I've spent about an hour trying to figure out how to swap floppy images. Best I got out of chatgpt was to open a second session with dosemu -t and filly with $_floppy but it doesn't seem to make much sense and didn't work anyway.

Any suggestions?

Thanks,

**Comment by stsp** [comment]
Any idea why chatgpt didn't allow
you to just replace the floppy file
by hands, or use symlink and change
it?

**Reply by andrewbird** [reply]
So just to expand on this, this is how I do it using two terminal windows:

##### Term 1
- configure dosemu to use the file ~/.dosemu/floppy.img as the floppy image

##### Term 2
- cp ~/Disk1.img ~/.dosemu/floppy.img

##### Term 1
- start dosemu
- run installation program until prompt for second disk

##### Term 2
- cp ~/Disk2.img ~/.dosemu/floppy.img

##### Term 1
- continue installation until prompt for third disk

##### Term 2
- cp ~/Disk3.img ~/.dosemu/floppy.img

##### Term 1
- continue installation until prompt for fourth disk

##### Term 2
- cp ~/Disk4.img ~/.dosemu/floppy.img

##### Term 1
- continue installation until finish

**Reply by stsp** [reply]
I think you'd better add `--remove-destination`
to `cp`, as otherwise it overwrites an existing
file, which may still be opened by dosemu2.
Also if dosemu2 doesn't close the image file
(due to some bug or a descriptor leak), this
way you'll make the problem visible.

**Reply by andrewbird** [reply]
> as otherwise it overwrites an existing
file, which may still be opened by dosemu2.

Actually I was relying upon that very behaviour, because if the file is open and then it is replaced (by delete, then create), dosemu will still serve up the original content to DOS if read from the original fd.

**Reply by stsp** [reply]
That behaviour is very dangerous,
and if dosemu2 forgets to close the
image, its a bug that should be fixed,
not "relied upon".

**Reply by andrewbird** [reply]
So do you expect that every dos floppy read should make dosemu do open/read/close of the image?

On a second note, I was just going to test this out, but dosemu doesn't seem to be able to open a floppy image. My setup may be awkward though I have an NFS home.
~~~
ajb@calypso:~/Desktop$ dosemu
ERROR: can't stat /home/ajb/.dosemu/a.img
ERROR: dev /home/ajb/.dosemu/a.img is wrong type
~~~
~~~
ajb@calypso:~/Desktop$ file /home/ajb/.dosemu/a.img
/home/ajb/.dosemu/a.img: DOS/MBR boot sector, code offset 0x3c+2, OEM-ID "mkfs.fat", root entries 224, sectors 2880 (volumes <=32 MB), sectors/FAT 9, sectors/track 18, serial number 0x81bfc7e, unlabeled, FAT (12 bit), followed by FAT
~~~

created by
~~~
ajb@calypso:~/.dosemu$ dd if=/dev/zero of=a.img bs=1024 count=1440
1440+0 records in
1440+0 records out
1474560 bytes (1.5 MB, 1.4 MiB) copied, 0.0554955 s, 26.6 MB/s
ajb@calypso:~/.dosemu$ mkfs -t vfat a.img
mkfs.fat 4.2 (2021-01-31)
~~~

Shall I open an issue, or is it something obviously wrong that I'm doing?

**Reply by andrewbird** [reply]
A little more info, I moved the image to /tmp and changed the dosemurc to reference it and then dosemu was able to start up and access it. I figure that the problem is something to do with my NFS home area.

**Comment by nelgin** [comment]
I ended up mounting the .img in Linux and copying all the files into one directory then mounting that in dosemu since there were no clashing filenames.

**Comment by stsp** [comment]
No, it likely needs -no-priv-sep just as everything else now. :(

Пятница, 19 июля 2024 г получено от Andrew Bird:

**Reply by andrewbird** [reply]
Yes, that was it, thanks.

**Comment by nelgin** [comment]
So I should be running with -no-priv-sep ?

**Reply by stsp** [reply]
No, I'll fix floppy.
But you can try `-no-priv-sep` once
something breaks, as this is an all-breaker.

**Reply by stsp** [reply]
There were not so many all-breakers,
as far as I can recall. Early versions of
fdpp or comcom32 were not all-breakers,
as they were not defaulted for quite long.
So I guess this is actually the first one. :(

**Comment by stsp** [comment]
Images fixed.

**Reply by andrewbird** [reply]
Thanks, floppy images now work for me on NFS home without needing to use -no-priv-sep.

**Reply by stsp** [reply]
Hope you already verified that they
correctly close, too.

**Reply by andrewbird** [reply]
I just about to do so, but I got another error with floppy access
~~~
FreeCom version 0.84-pre2 XMS_Swap [Aug 28 2006 00:29:00]
BLASTER=A220 I5 D1 H5 P300 T6
MIDI=SYNTH:2 MAP:E MODE:0
Welcome to dosemu2!
    Build 2.0pre9
Interactive Prompt!
C:\>dir a:
reject /home/ajb/.dosemu/a.img -1
ERROR: fssvc_open returned Operation not permitted, exiting
ERROR: Please update dosemu from git, compile it with debug
info and make a bug report with the content of ~/.dosemu/boot.log at
https://github.com/dosemu2/dosemu2/issues
Please provide any additional info you can, like the test-cases,
URLs and all the rest that fits.
~~~

**Reply by stsp** [reply]
So it still works with `-no-priv-sep`?
How to reproduce?

**Reply by andrewbird** [reply]
Actually no, just the same result if I add -no-priv-sep

Also the same effect if I move the floppy image to /tmp (and reconfigure dosemurc)

To reproduce:
- Add a formatted floppy image
- configure dosemurc
- start dosemu
- type dir a:
- see dosemu complain and exit

**Reply by andrewbird** [reply]
Also I am using FreeCOM, not sure if that's significant.

**Reply by stsp** [reply]
What is `configure dosemurc`?

**Reply by andrewbird** [reply]
`vi ~/.dosemu/dosemurc` and add `$_floppy_a = "a.img"`

**Reply by stsp** [reply]
Ahha, I had to comment out `$_lredir_paths`,
otherwise it doesn't break!

**Reply by stsp** [reply]
OK, it doesn't breaks if `$_lredir_paths`
has `~` in a list. :)

**Reply by stsp** [reply]
I hope its now fixed.

**Reply by andrewbird** [reply]
Thanks I can now do the 'dir a:' without dosemu exiting. I'll try later regarding the open/read/close fd lifecycle.
